### PR TITLE
Fix log matplot figure bug

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## What changes are proposed in this pull request?

Fix log matplot figure bug https://github.com/mlflow/mlflow/issues/5404

## How is this patch tested?

Manually

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @WeichenXu123 !

Can we add a test?

**Comment by harupy** [comment]
We have a test for `mlflow.log_figure`, we run `import matplotlib.pyplot` which makes `matplotlib.figure` available:

https://github.com/mlflow/mlflow/blob/011ee78cebd492ab3d75d8f87f5c9d80334d2bf6/tests/tracking/test_tracking.py#L669-L677

File: mlflow/tracking/fluent.py [hunk 1]
```diff
@@ -52,6 +52,7 @@
 if TYPE_CHECKING:
     import pandas  # pylint: disable=unused-import
     import matplotlib  # pylint: disable=unused-import
+    import matplotlib.figure
```
Comment (harupy) [inline mlflow/tracking/fluent.py:55]: why do we need this import statement?

**Comment by harupy** [comment]
@ozymand1a Do you have code that can reproduce the issue?

File: mlflow/tracking/fluent.py (hunk 1 above)
Comment (WeichenXu123) [reply mlflow/tracking/fluent.py:55]: I saw the type definition using it: e.g. `figure: Union["matplotlib.figure.Figure"]...`

Comment (harupy) [reply mlflow/tracking/fluent.py:55]: I see, makes sense!

**Comment by dbczumar** [comment]
> We have a test for `mlflow.log_figure`. In this test, we run `import matplotlib.pyplot` which makes `matplotlib.figure` available:
> 
> https://github.com/mlflow/mlflow/blob/011ee78cebd492ab3d75d8f87f5c9d80334d2bf6/tests/tracking/test_tracking.py#L669-L677

I see. Can we remove `import matplotlib.pyplot` from this test case?

**Comment by harupy** [comment]
@dbczumar We can't because we use `matplotlib.pyplot` to create a plot.

**Comment by dbczumar** [comment]
> @dbczumar We can't because we use `matplotlib.pyplot` to create a plot.

I see. In that case, I think it's fine to just merge the fix as-is.

File: mlflow/tracking/fluent.py (hunk 1 above)
Comment (harupy) [reply mlflow/tracking/fluent.py:55]: Do we need `# pylint: disable=unused-import` here?

**Comment by harupy** [comment]
I agree, lgtm!

**Review by harupy — APPROVED** [review]

**Comment by ozymand1a** [comment]
> @ozymand1a Do you have code that can reproduce the issue?

image

**Comment by harupy** [comment]
@ozymand1a You encountered this issue while using MLflow, right? Could you share your code that contains mlflow and caused the issue?

**Comment by ozymand1a** [comment]
I use this function:

```python
@rank_zero_only
def log_figure(
        loggers: list,
        run_id: str,
        figure: Union["matplotlib.figure.Figure", "plotly.graph_objects.Figure"],
        artifact_file: str
) -> None:
    from mlflow.tracking.client import MlflowClient
    for logger in loggers:
        if logger.__class__ == MlflowClient:
            logger: MlflowClient = logger
            logger.log_figure(run_id, figure, artifact_file)
        else:
            raise RuntimeError(f"Not implemented <log_artifact> method for {logger.__class__} logger!")
```

and call it at pytorch lightning module this way:
```python
    def test_epoch_end(self, outputs) -> None:

        ...

        fig = plot_confusion_matrix(
            cm=cf_matrix,
            labels=labels,
            title='Confusion_matrix'
        )

        if 'mlflow' in self.cfg.logger:
            log_figure(self.logger.experiment, self.logger.version, fig, f'confusion_matrix.html')
```

**Comment by harupy** [comment]
@ozymand1a Can you share the full code?

**Comment by ozymand1a** [comment]
> @ozymand1a Can you share the entire code, not the parts of the code?
> 
> I'm curious:
> 
> * what `plot_confusion_matrix` returns.
> * which modules your code imports.

@harupy  `plot_confusion_matrix` returns `plotly.graph_objects.Figure` object

```python
def plot_confusion_matrix(
        cm: np.array,
        labels: List[str],
        title: str = 'Confusion Matrix'
):
    """
    cm : confusion matrix list(list)
    labels : name of the data list(str)
    title : title for the heatmap
    """
    data = go.Heatmap(z=cm, y=labels, x=labels, colorscale='Portland')
    annotations = []
    for i, row in enumerate(cm):
        for j, value in enumerate(row):
            annotations.append(
                {
                    "x": labels[j],
                    "y": labels[i],
                    "font": {"color": "black"},
                    "text": str(value),
                    "xref": "x1",
                    "yref": "y1",
                    "showarrow": False
                }
            )
    layout = {
        "title": title,
        "title_x": 0.5,
        "xaxis": {"title": "Predicted value"},
        "yaxis": {"title": "Real value"},
        "annotations": annotations,
    }
    fig = go.Figure(data=data, layout=layout)
    fig['layout']['yaxis']['autorange'] = "reversed"
    return fig
```

I don't understand what do you call module imports. I mean is this mlflow, matplotlib versions, or something else?

**Comment by harupy** [comment]
```
import a
import b
import c
```

I think you import several modules / pakcage at the top of the script, right?

**Comment by ozymand1a** [comment]
@harupy Actually, I can not copy all code because of NDA. But I can describe code structure:

First, it is a logging script:
```python3
import numpy as np
from typing import Union
from pytorch_lightning import LightningModule
from pytorch_lightning.utilities import rank_zero_only

@rank_zero_only
def log_figure(
        loggers: list,
        run_id: str,
        figure: Union["matplotlib.figure.Figure", "plotly.graph_objects.Figure"],
        artifact_file: str
) -> None:
    from mlflow.tracking.client import MlflowClient
    for logger in loggers:
        if logger.__class__ == MlflowClient:
            logger: MlflowClient = logger
            logger.log_figure(run_id, figure, artifact_file)
        else:
            raise RuntimeError(f"Not implemented <log_artifact> method for {logger.__class__} logger!")
```
Second I have a basic pytorch lightning module like from this repo: https://github.com/ashleve/lightning-hydra-template
My imports:
```python3
import hydra
from omegaconf import DictConfig
from sklearn.metrics import confusion_matrix

import torch
from pytorch_lightning import LightningModule
from torchmetrics import MetricCollection
```

I imported matplotlib of the newest version at lightning module but it doesn't help me.

**Pull request merged**